configuration.ml

```diff
@@ -135,3 +136,66 @@ let of_json json =
                     current_version))
       | Some _ -> Error (`Msg "configuration: version must be an integer"))
   | _ -> Error (`Msg "configuration: expected a dictionary")
+
+let of_json json =
+  match json with
+  | `Assoc xs -> (
+      match get "version" xs with
+      | None -> Error (`Msg "configuration: couldn't find a version")
+      | Some (`Int v) ->
+          if v = current_version then
+            match
+              ( get "name" xs,
+                get "certificate" xs,
+                get "private_key" xs,
+                get "server_ip" xs,
+                get "server_port" xs,
+                get "updated_at" xs )
+            with
+            | ( Some (`String name),
+                Some (`String cert),
+                Some (`String key),
+                Some (`String server_ip),
+                Some (`Int server_port),
+                Some (`String updated_at) ) ->
+                let ( let* ) = Result.bind in
+                let* certificate = X509.Certificate.decode_pem cert in
+                let* private_key = X509.Private_key.decode_pem key in
+                let* () =
+                  if
+                    not
+                      (String.equal
+                         (X509.Public_key.fingerprint
+                            (X509.Certificate.public_key certificate))
+                         (X509.Public_key.fingerprint
+                            (X509.Private_key.public private_key)))
+                  then Error (`Msg "certificate and private key do not match")
+                  else Ok ()
+                in
+                let* server_ip = Ipaddr.of_string server_ip in
+                let* updated_at = Utils.TimeHelper.ptime_of_string updated_at in
+
+                Ok
+                  {
+                    name;
+                    certificate;
+                    private_key;
+                    server_ip;
+                    server_port;
+                    updated_at;
+                  }
+            | _ ->
+                Error
+                  (`Msg
+                     (Fmt.str "configuration: unexpected types, got %s"
+                        (Yojson.Basic.to_string (`Assoc xs))))
+          else
```

Comment:
here's our v1 configurations that we have on disk, and we should properly parse them (and add the `"default"` name as well -- we have to ensure there's only a single configuration, though).